### PR TITLE
[node-core-library] Replace IFileModeBits with a more flexible PosixModeBits enum

### PR DESCRIPTION
**Pull request by pgonzal**
This PR fixes several issues encountered with the IFileModeBits interface that was recently added for the changePermissionBits() API:

- There was no easy way to read or test the bits of a file
- The FileModeBits members used InitialCaps instead of camelCase which doesn't follow our standard
- The getStatistics() API returns fs.Stats with a "mode" member, but there weren't helpers for converting between IFileModeBits and the octal number
- The APIs inconsistently used the word "mode" and "getPermissionBits" for variables of type IFileModeBits

This PR replaces the IFileModeBits  interface with a PosixModeBits enum that can be used interchangeably with the NodeJS "mode" integer from e.g. `fs.Stats`.

### CODE REVIEW

File: libraries/node-core-library/src/FileSystem.ts [hunk 1]
```diff
@@ -217,17 +283,49 @@ export class FileSystem {
    * Changes the permissions (i.e. file mode bits) for a filesystem object.
    * Behind the scenes it uses `fs.chmodSync()`.
    * @param path - The absolute or relative path to the object that should be updated.
+   * @param modeBits - UNIX-style file mode bits (e.g. 777 or 666 etc)
```
Comment (nickpape-msft) [inline libraries/node-core-library/src/FileSystem.ts]: >odeBits - UNIX-style file mode bits (e.g. 777 or 666 etc) [](start = 13, length = 57)

change this comment

File: libraries/node-core-library/src/FileSystem.ts [hunk 2]
```diff
@@ -217,17 +283,49 @@ export class FileSystem {
...
+   */
+  public static changePosixModeBits(path: string, mode: PosixModeBits): void {
+    fs.chmodSync(path, mode);
+  }
+
+  /**
+   * Retrieves the permissions (i.e. file mode bits) for a filesystem object.
+   * Behind the scenes it uses `fs.chmodSync()`.
+   * @param path - The absolute or relative path to the object that should be updated.
    * @param mode - UNIX-style file mode bits (e.g. 777 or 666 etc)
```
Comment (nickpape-msft) [inline libraries/node-core-library/src/FileSystem.ts]: >UNIX-style file mode bits (e.g. 777 or 666 etc) [](start = 19, length = 47)

update to reflect that we are referencing your special enum

File: libraries/node-core-library/src/FileSystem.ts [hunk 3]
```diff
@@ -217,17 +283,49 @@ export class FileSystem {
...
+    result += (mode & PosixModeBits.UserWrite) ? 'w' : '-';
+    result += (mode & PosixModeBits.UserExecute) ? 'x' : '-';
+
+    result += (mode & PosixModeBits.GroupRead) ? 'r' : '-';
+    result += (mode & PosixModeBits.GroupWrite) ? 'w' : '-';
+    result += (mode & PosixModeBits.GroupExecute) ? 'x' : '-';
+
+    result += (mode & PosixModeBits.OthersRead) ? 'r' : '-';
+    result += (mode & PosixModeBits.OthersWrite) ? 'w' : '-';
+    result += (mode & PosixModeBits.OthersExecute) ? 'x' : '-';
```
Comment (nickpape-msft) [inline libraries/node-core-library/src/FileSystem.ts]: I think this operation is slower than setting up an array and using .join or reduce...

Comment (pgonzal) [reply libraries/node-core-library/src/FileSystem.ts]: The strings are tiny

---
In reply to: [205244478](https://github.com/Microsoft/web-build-tools/pull/754#discussion_r205244478) [](ancestors = 205244478)

**Pull request merged**